File: python/test_dr_companion.py
"""dr_companion.py against the real app, not a stand-in for it.

Everything below connects to a genuinely running DR Companion - reading its
actual `script-api.port`/`script-api.token` files, the same way any real
script would, rather than a mock server that only proves the two sides agree
with themselves. `game_link.rs`'s own tests use a bare TcpListener standing in
for Lich, which is the right call there because Lich is not this project's
code to run in a test; the script API is, and the app was already open when
this was written, so there is no excuse for testing anything less than the
real thing.

Skips loudly rather than passing quietly when the app is not running - see
`_require_running_app`. A green run against nothing would be indistinguishable
from a green run that actually exercised the socket, which is exactly the
failure this project's own working notes call out repeatedly: a check that
cannot fail is not a check.

Run with the app open:

    python python/test_dr_companion.py
"""

from __future__ import annotations

import sys
import threading
import time

from dr_companion import Companion, ConnectionError, Line, Status, _data_dir


def _require_running_app() -> None:
    data_dir = _data_dir()
    port_file = data_dir / "script-api.port"
    token_file = data_dir / "script-api.token"
    if not port_file.exists() or not token_file.exists():
        print(
            f"SKIPPED: no script-api files in {data_dir} - is DR Companion running? "
            "This is not a pass; nothing was checked."
        )
        sys.exit(2)


def _ok(label: str, cond: bool, detail: str = "") -> bool:
    print(f"{'OK  ' if cond else 'FAIL'} {label:<58}{detail}")
    return cond


def main() -> int:
    _require_running_app()
    failed = 0

    print("-- status() and send() connect on their own, like run() does --")
    # Found by running python/examples/hello.py, which calls status() before
    # run() gets a chance to connect: status() raised NotConnected. A method
    # a script would reasonably call first should not be the one exception
    # to "you do not have to call connect() yourself".
    for label, action in [
        ("status()", lambda comp: comp.status()),
        ("send()", lambda comp: comp.send("--dr-companion-python-api-smoke-test--")),
    ]:
        fresh = Companion()
        try:
            action(fresh)
            ok = fresh._sock is not None
        except Exception as e:  # noqa: BLE001 - reporting, not handling
            ok = False
            print(f"     {label} raised: {e}")
        if not _ok(f"{label} connects without connect() being called first", ok):
            failed += 1
        fresh.close()

    print("\n-- connecting to the real app --")
    c = Companion()
    c.connect()
    if not _ok("connected and authenticated", True):
        failed += 1

    print("\n-- status works even with nothing attached --")
    st = c.status()
    if not _ok("status came back as the right type", isinstance(st, Status), repr(st)):
        failed += 1
    # Not asserting `connected` either way - whether Lich is attached is
    # this machine's real state right now, not something this test controls,
    # and asserting a specific value would make the test depend on a human
    # having clicked Attach. What matters is that a real, well-formed answer
    # came back at all.
    if not _ok("note is a string", isinstance(st.note, str), repr(st.note)):
        failed += 1

    print("\n-- a bad token is refused, against the real server --")
    try:
        bad = Companion(token="not-the-real-token")
        bad.connect()
        _ok("a wrong token is rejected", False, "connected anyway")
        failed += 1
    except ConnectionError as e:
        _ok("a wrong token is rejected", True, str(e)[:60])

    print("\n-- lines arrive as they are broadcast, not just in principle --")
    # Send a command. Whether or not a game is attached, sending itself either
    # succeeds (attached) or comes back as a documented error (not attached) -
    # both are the real server answering honestly, and this proves `send`
    # reaches the app rather than only proving the client can format a string.
    c.send("--dr-companion-python-api-smoke-test--")
    got_error = threading.Event()
    error_text = []

    original_dispatch = c._dispatch

    def watching_dispatch(msg):
        if msg.get("type") == "error":
            error_text.append(msg.get("message", ""))
            got_error.set()
        original_dispatch(msg)

    c._dispatch = watching_dispatch

    # Give the app a moment to answer if it is going to (only fires when not
    # attached - a real attach would just send the text into the game, which
    # is not an error and produces no reply to wait for).
    deadline = time.time() + 1.5
    while time.time() < deadline and not got_error.is_set():
        c._sock.settimeout(0.2)
        try:
            msg = c._read_message()
        except OSError:
            break
        if msg is not None:
            watching_dispatch(msg)
        else:
            break
    c._sock.settimeout(None)

    if st.connected:
        print("OK   (Lich is attached right now - send() had a real effect, not asserted here)")
    else:
        if not _ok(
            "not attached: send() is answered with an error, not silence",
            got_error.is_set(),
            error_text[0] if error_text else "(nothing came back)",
        ):
            failed += 1

    c.close()

    print("\n-- the second connection gets a fresh, working socket too --")
    c2 = Companion()
    c2.connect()
    st2 = c2.status()
    if not _ok("a second real connection also authenticates and answers", isinstance(st2, Status)):
        failed += 1
    c2.close()

    print(f"\n{failed} failed" if failed else "\nall passed")
    return 1 if failed else 0


if __name__ == "__main__":
    raise SystemExit(main())

File: typescript/test_dr_companion.ts
/**
 * dr_companion.ts against a bare TCP server standing in for the app -
 * the same pattern `game_link.rs`'s own Rust tests use for Lich (a bare
 * `TcpListener`), and the honest option here: this project's Tauri app does
 * not run in this environment, so there is no live DR Companion to test
 * against the way `python/test_dr_companion.py` does. A fake server that
 * speaks the real wire protocol is a closer test of this file than skipping
 * the network entirely.
 *
 * Run with:
 *
 *     node --experimental-strip-types typescript/test_dr_companion.ts
 *     node typescript/test_dr_companion.ts   # Node 24+, no flag needed
 */

import { createServer, type Socket } from 'node:net'
import { Companion, type Line, type Status } from './dr_companion.ts'

let failed = 0
function ok(label: string, cond: boolean, detail = ''): void {
  console.log(`${cond ? 'OK  ' : 'FAIL'} ${label.padEnd(58)}${detail}`)
  if (!cond) failed++
}

/** A minimal fake app: hello, auth, then whatever the test wants to push. */
function startFakeServer(token: string): Promise<{ port: number; close: () => void; send: (obj: object) => void }> {
  return new Promise((resolve) => {
    let client: Socket | null = null
    const server = createServer((socket) => {
      client = socket
      let buf = ''
      socket.write(JSON.stringify({ type: 'hello', version: 1 }) + '\n')
      socket.on('data', (chunk) => {
        buf += chunk.toString('utf8')
        const nl = buf.indexOf('\n')
        if (nl === -1) return
        const msg = JSON.parse(buf.slice(0, nl))
        buf = buf.slice(nl + 1)
        if (msg.type === 'auth') {
          if (msg.token === token) {
            socket.write(JSON.stringify({ type: 'auth_ok' }) + '\n')
          } else {
            socket.write(JSON.stringify({ type: 'auth_failed' }) + '\n')
            socket.end()
          }
        }
        // 'send'/'status' requests from the client are otherwise ignored -
        // this fake only pushes what the test explicitly asks it to.
      })
    })
    server.listen(0, '127.0.0.1', () => {
      const addr = server.address()
      if (addr === null || typeof addr === 'string') throw new Error('unexpected server address')
      resolve({
        port: addr.port,
        close: () => server.close(),
        send: (obj: object) => {
          if (!client) throw new Error('no client connected yet')
          client.write(JSON.stringify(obj) + '\n')
        },
      })
    })
  })
}

async function main(): Promise<number> {
  const token = 'test-token-12345'

  // -- connect() succeeds against a real handshake -----------------------
  {
    const fake = await startFakeServer(token)
    const c = new Companion({ host: '127.0.0.1', port: fake.port, token })
    await c.connect()
    ok('connect() completes the hello/auth handshake', true)
    c.close()
    fake.close()
  }

  // -- connect() rejects a bad token ---------------------------------
  {
    const fake = await startFakeServer(token)
    const c = new Companion({ host: '127.0.0.1', port: fake.port, token: 'wrong' })
    let threw = false
    try {
      await c.connect()
    } catch {
      threw = true
    }
    ok('connect() rejects a bad token', threw)
    fake.close()
  }

  // -- line/state events fire, in order, after connect --------------------
  {
    const fake = await startFakeServer(token)
    const c = new Companion({ host: '127.0.0.1', port: fake.port, token })
    await c.connect()

    const lines: Line[] = []
    const states: Status[] = []
    c.on('line', (l) => lines.push(l))
    c.on('state', (s) => states.push(s))

    fake.send({ type: 'line', seq: 1, text: 'hello world' })
    fake.send({ type: 'state', connected: true, host: '127.0.0.1', port: 11024, lines: 1, note: '' })

    await new Promise((r) => setTimeout(r, 50))

    ok('a line event fires with the right text', lines.length === 1 && lines[0].text === 'hello world', JSON.stringify(lines))
    ok('a state event fires with the right shape', states.length === 1 && states[0].connected === true, JSON.stringify(states))

    c.close()
    fake.close()
  }

  // -- status() resolves on the next state message, not before ------------
  {
    const fake = await startFakeServer(token)
    const c = new Companion({ host: '127.0.0.1', port: fake.port, token })
    await c.connect()

    // A line arriving before the state reply must still be emitted, not
    // swallowed by the pending status() call - the same guarantee
    // dr_companion.py's status() makes.
    const lines: Line[] = []
    c.on('line', (l) => lines.push(l))

    const statusPromise = c.status()
    fake.send({ type: 'line', seq: 2, text: 'a line that arrived first' })
    fake.send({ type: 'state', connected: true, host: '127.0.0.1', port: 11024, lines: 2, note: '' })

    const status = await statusPromise
    ok('status() resolves with the state message', status.lines === 2, JSON.stringify(status))
    ok("status() does not swallow a line that arrived first", lines.length === 1, JSON.stringify(lines))

    c.close()
    fake.close()
  }

  // -- a malformed message on the wire is reported, not thrown -----------
  {
    const fake = await startFakeServer(token)
    const c = new Companion({ host: '127.0.0.1', port: fake.port, token })
    await c.connect()

    const errors: string[] = []
    c.on('error', (msg) => errors.push(msg))
    fake.send({ type: 'error', message: 'something went wrong' })
    await new Promise((r) => setTimeout(r, 50))
    ok("an 'error' message is emitted as an 'error' event", errors.length === 1 && errors[0] === 'something went wrong')

    c.close()
    fake.close()
  }

  if (failed) {
    console.log(`\n${failed} check(s) FAILED`)
    return 1
  }
  console.log('\nall checks OK')
  return 0
}

main().then((code) => process.exit(code))
